src/analysis/history_utils.hpp

```diff
@@ -48,9 +50,116 @@ KOKKOS_INLINE_FUNCTION Real CalcMassFlux(Pack &pack, Geometry &geom, const int p
   const Real W = phoebus::GetLorentzFactor(vel, gcov4);
   const Real ucon = vel[0] - shift[0] * W / lapse;
 
-  return lapse * gdet * pack(b, prho, k, j, i) * ucon;
+  return -lapse * gdet * pack(b, prho, k, j, i) * ucon;
 }
 
+template <typename Pack, typename Geometry>
+KOKKOS_INLINE_FUNCTION Real CalcMagnetization(Pack &pack, Geometry &geom, const int pb_lo,
+                                              const int pb_hi, const int prho,
+                                              const int b, const int k, const int j,
+                                              const int i) {
+  Real gam[3][3];
+  geom.Metric(CellLocation::Cent, k, j, i, gam);
+  const Real Bp[] = {pack(b, pb_lo, k, j, i), pack(b, pb_lo + 1, k, j, i),
+                     pack(b, pb_hi, k, j, i)};
+  const Real rho = pack(b, prho, k, j, i);
+
+  Real Bsq = 0;
+  for (int m = 0; m < 3; m++)
+    for (int n = 0; n < 3; n++) {
+      Bsq += gam[m][n] * Bp[m] * Bp[n];
+    }
+
+  return Bsq / rho;
+}
+
+template <typename Pack, typename Geometry>
+KOKKOS_INLINE_FUNCTION Real CalcEMEnergyFlux(Pack &pack, Geometry &geom,
+                                             const int pvel_lo, const int pvel_hi,
+                                             const int pb_lo, const int pb_hi,
+                                             const int b, const int k, const int j,
+                                             const int i) {
+  Real gam[3][3];
+  geom.Metric(CellLocation::Cent, k, j, i, gam);
+  Real gdet = geom.DetGamma(CellLocation::Cent, k, j, i);
+  Real lapse = geom.Lapse(CellLocation::Cent, k, j, i);
+
+  const Real vcon[] = {pack(b, pvel_lo, k, j, i), pack(b, pvel_lo + 1, k, j, i),
+                       pack(b, pvel_hi, k, j, i)};
+  const Real Bp[] = {pack(b, pb_lo, k, j, i), pack(b, pb_lo + 1, k, j, i),
+                     pack(b, pb_hi, k, j, i)};
+  const Real W = phoebus::GetLorentzFactor(vcon, gam);
+
+  Real Bsq = 0.0;
+  Real Bdotv = 0.0;
+  Real vcov[3] = {0};
+  Real Bcov[3] = {0};
+  for (int m = 0; m < 3; m++)
+    for (int n = 0; n < 3; n++) {
+      Bsq += gam[m][n] * Bp[m] * Bp[n];
+      vcov[m] += gam[m][n] * vcon[n];
+      Bcov[m] += gam[m][n] * Bp[n];
+      Bdotv += gam[m][n] * Bp[m] * robust::ratio(vcon[n], W);
+    }
+
+  const Real b0 = Bdotv * W / lapse;
+  const Real ucon0 = W / lapse;
+  const Real ucov1 = vcov[0];
+  const Real bcov1 = robust::ratio((Bcov[0] + lapse * b0 * ucov1), W);
+
+  const Real bsq = phoebus::GetMagneticFieldSquared(gam, vcon, Bp, W);
+
+  // energy flux of EM stress energy tensor
+  const Real T_EM_01 = bsq * ucon0 * ucov1 - b0 * bcov1;
+
+  return +T_EM_01 * lapse * gdet;
```

Comment:
Don't we want T^1_0 not T^0_1?

Reply:
Ah, thanks for this. I believe I've pushed the necessary changes.